scalaz-test/src/main/scala/ztest/data.scala

```diff
@@ -0,0 +1,30 @@
+package scalaz.test
+
+import scala._, scala.Predef._
+
+import scalaz.Scalaz._, scalaz.core.EqClass
+
+trait Harness[F[_], G[_]] {
+  def section[A](annotation: String)(test: G[A]): G[A]
+  def shared[A](fa: F[A]): G[A]
+  def test(assertion: F[TestResult]): G[Unit]
+  final def namedTest(annotation: String)(errs: F[TestResult]): G[Unit] =
+    section(annotation)(test(errs))
+}
+
+sealed trait TestResult
+final class Thrown(val thrown: Throwable) extends TestResult
+object Thrown {
+  def apply(thrown: Throwable): TestResult = new Thrown(thrown)
+}
+object Success extends TestResult {
+  def apply(): TestResult = this
+}
+object Failure extends TestResult {
+  def apply(): TestResult = this
+}
+object TestResult {
+  implicit val equalTestResult: Eq[TestResult] = instanceOf(new EqClass[TestResult] {
+    def equal(first: TestResult, second: TestResult) = first == second
```

Comment:
So two `Thrown`s are never equal?

Reply:
Might as well treat them as data and compare them based on `cause`, `stackTrace`, `message`, and `getClass`.

Reply:
The only sane approach (that I know) to equality on open types is to treat them as equal if they are observationally equivalent in terms of their super type methods.
And unless we find a better approach I suggest this as another principle for Scalazzi code

Reply:
`Throwable` is indeed a nasty one....

Reply:
@hrhino Not treating any throwables as equal, no.

Reply:
I'm okay with that.